extensions/csv/src/main/java/io/deephaven/csv/reading/CellGrabber.java

```diff
@@ -0,0 +1,276 @@
+                if (!RangeTests.isWhitespace(ch)) {
+                    return;
+                }
+                ++offset;
+            }
+        } while (tryEnsureMore());
+    }
+
+    /**
+     * @return true iff the field is the last in the row, false otherwise.
+     */
+    private boolean finishField(final CharSlice dest) throws CsvReaderException {
+        while (true) {
+            while (offset != size) {
+                final char ch = buffer[offset];
+                if (ch == fieldDelimiter) {
+                    finish(dest);
+                    ++offset;
+                    return false;
+                }
+                if (ch == '\n') {
+                    finish(dest);
+                    ++offset;
+                    return true;
+                }
+                if (ch == '\r') {
+                    finish(dest);
+                    ++offset;
+                    if (tryEnsureMore()) {
+                        // might be \r\n
+                        if (buffer[offset] == '\n') {
+                            ++offset;
+                        }
+                    }
+                    return true;
+                }
+                ++offset;
+            }
+            if (!tryEnsureMore()) {
+                finish(dest);
+                // End of file sets last in row.
+                return true;
+            }
+        }
+    }
+
+    /**
+     * @return true if there are more characters.
+     */
+    private boolean tryEnsureMore() throws CsvReaderException {
+        if (offset != size) {
+            return true;
+        }
+        spillRange();
+        refillBuffer();
+        return size != 0;
+    }
+
+    private void spillRange() {
+        spillBuffer.append(buffer, startOffset, offset - startOffset);
+        startOffset = offset;
+    }
+
+    private void refillBuffer() throws CsvReaderException {
+        offset = 0;
+        startOffset = 0;
+        try {
+            while (true) {
+                final int charsRead = reader.read(buffer);
+                if (charsRead < 0) {
+                    size = 0;
+                    return;
+                }
+                if (charsRead > 0) {
+                    size = charsRead;
+                    return;
+                }
+                // I don't think zero-length reads happen, but if they do, I'll just keep trying.
+            }
+        } catch (IOException inner) {
+            throw new CsvReaderException("Caught exception", inner);
+        }
+    }
+
+    private void finish(final CharSlice dest) {
+        if (spillBuffer.size() == 0) {
+            // If we never spilled then our whole output is in the input buffer. So we can
+            // just return a slice of the input buffer.
+            dest.reset(buffer, startOffset, offset);
+            return;
+        }
+        // Otherwise, append we need to append whatever residual is left to spillBuffer
+        // and return a slice of spillBuffer.
```

Comment:
Btw you have the best use of comments to explain code flow that I've ever seen.  Meaning, the combination of code clarity and meaningful variable names with tactically placed comments to guide the reader.  It is really great, kudos, I recognize it because I know I am not particularly good at this.  In terms of code itselfI don'\t know what is your particular way to reach this level of clarity, but for me to get here means doing several passes of write and re-write.  Anyway, whatever it is, keep doing it, is great.

Reply:
Thank you very much. I really appreciate your comment and am honored by it 🌙 🥧 💯 🎁